drivers/bicker_ser.c

```diff
@@ -0,0 +1,379 @@
+	DRIVER_NAME,
+	DRIVER_VERSION,
+	"Nicola Fontana <ntd@entidi.it>",
+	DRV_EXPERIMENTAL,
+	{ NULL }
+};
+
+static ssize_t bicker_send(char cmd, const void *data, size_t datalen)
+{
+	char buf[BICKER_PACKET];
+	size_t buflen;
+	ssize_t ret;
+
+	ser_flush_io(upsfd);
+
+	if (data != NULL) {
+		if (datalen > 252) {
+			upslogx(LOG_ERR, "Data size exceeded: %d > 252",
+				(int)datalen);
+			return 0;
+		}
+		memcpy(buf + 4, data, datalen);
+	} else {
+		datalen = 0;
+	}
+
+	buf[0] = BICKER_SOH;
+	buf[1] = datalen + 3; /* Packet size must include the header */
+	buf[2] = '\x03'; /* Command index is always 3 */
+	buf[3] = cmd;
+	buf[4 + datalen] = BICKER_EOT;
+
+	/* The full packet includes SOH and EOT bytes too */
+	buflen = datalen + 3 + 2;
+
+	ret = ser_send_buf(upsfd, buf, buflen);
+	if (ret < 0) {
+		upslog_with_errno(LOG_WARNING, "ser_send_buf failed");
+		return ret;
+	} else if ((size_t) ret != buflen) {
+		upslogx(LOG_WARNING, "ser_send_buf returned %d instead of %d",
+			(int)ret, (int)buflen);
+		return 0;
+	}
+
+	upsdebug_hex(3, "bicker_send", buf, buflen);
+	return ret;
+}
+
+static ssize_t bicker_receive_buf(void *dst, size_t dstlen)
+{
+	ssize_t ret;
+
+	ret = ser_get_buf(upsfd, dst, dstlen, 1, 1);
+	if (ret < 0) {
+		upslog_with_errno(LOG_WARNING, "ser_get_buf failed");
+		return ret;
+	} else if ((size_t) ret != dstlen) {
+		upslogx(LOG_WARNING, "ser_get_buf returned %d instead of %d",
+			(int)ret, (int)dstlen);
+		return 0;
+	}
+
+	upsdebug_hex(3, "bicker_receive_buf", dst, dstlen);
+	return ret;
+}
+
+static ssize_t bicker_receive(char cmd, void *dst, size_t dstlen)
+{
+	ssize_t ret;
+	size_t datalen;
+	char buf[BICKER_PACKET];
+
+	ret = bicker_receive_buf(buf, 1);
+	if (ret <= 0) {
+		return ret;
+	} else if (buf[0] != BICKER_SOH) {
+		upslogx(LOG_WARNING, "Received 0x%02X instead of SOH (0x%02X)",
+			(unsigned)buf[0], (unsigned)BICKER_SOH);
+		return 0;
+	}
+
+	ret = bicker_receive_buf(buf + 1, 1);
+	if (ret <= 0) {
+		return ret;
+	}
+
+	datalen = buf[1] - 3;
+
+	ret = bicker_receive_buf(buf + 2, datalen + 3);
+	if (ret <= 0) {
+		return ret;
+	} else if (buf[datalen + 4] != BICKER_EOT) {
+		upslogx(LOG_WARNING, "Received 0x%02X instead of EOT (0x%02X)",
+			(unsigned)buf[datalen + 4], (unsigned)BICKER_EOT);
+		return 0;
+	} else if (buf[3] != cmd) {
+		upslogx(LOG_WARNING, "Commands do not match: sent 0x%02X, received 0x%02X",
+			(unsigned)cmd, (unsigned)buf[1]);
+		return 0;
+	} else if (dstlen < datalen) {
+		upslogx(LOG_ERR, "Not enough space for the payload: %d < %d",
+			(int)dstlen, (int)datalen);
+		return 0;
+	}
+
+	if (dst != NULL) {
+		memcpy(dst, buf + 4, datalen);
+	}
+
+	upsdebug_hex(3, "bicker_receive", buf, datalen + 5);
+	return datalen;
+}
+
+static ssize_t bicker_read_int16(char cmd, int16_t *dst)
+{
+	ssize_t ret;
+
+	ret = bicker_send(cmd, NULL, 0);
+	if (ret <= 0) {
+		return ret;
+	}
+
+	/* XXX: by default, data is stored in little-endian so,
+	 * on big-endian platforms, a byte swap should be needed */
```

Comment:
Note we have the `AC_C_BIGENDIAN` macro called in `configure.ac` and later `AC_SUBST(WORDS_BIGENDIAN)`, since per https://www.gnu.org/software/autoconf/manual/autoconf-2.60/html_node/C-Compiler.html "The default for action-if-true is to define ‘WORDS_BIGENDIAN’. The default for action-if-false is to do nothing. And finally, the default for action-if-unknown is to abort configure and tell the installer which variable he should preset to bypass this test."

There is an example use-case in NUT libshut:
* https://github.com/networkupstools/nut/blob/master/drivers/libshut.c#L337-L350

Reply:
Thinking of it, if you know the endianness of the wire protocol words (which of the two bytes is significant), you can construct int's in driver code by bit-shifting one of the bytes by 8 bits (or multiplying one by 256) and get the 16-bit word in CPU-specific format. This pattern is actually often seen in NUT code.

Reply:
Should be fixed in 13ff8f52b8e22da673b43d5289c6fe41d3efdacd.